---
title: "Mapping eQTLs"
author: "Annique Claringbould"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
***
<br><br><br>
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Simple eQTL mapping

<br>In this tutorial, you will identify expression quantitative loci (eQTLs). You will be provided with a list of single nucleotide polymorphisms (SNPs). 

Let's start by cleaning our R environment and installing the necessary package: **MatrixEQTL**.  

```{r library, eval=FALSE}
rm(list = ls())
install.packages('MatrixEQTL')
library(MatrixEQTL)
```

<br><br>The package was developed by Andrey A. Shabalin, more information can be found here: http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/

<br><br>**Q1. Find out from the documentation on this package which models we can choose. What are the main differences between the models?**

<br><br>We start by setting the base directory, which should be the directory where you store the files from this tutorial: the genotypes, gene expression, SNP locations, and gene locations.

```{r base_dir, eval=FALSE}
base.dir ='[/your/directory/here/]'
```

<br><br>Then we set the model to linear:

```{r model, eval=FALSE}
useModel = modelLINEAR
```

<br><br>Set the file locations, so running the model becomes easy.

```{r data_locations, eval=FALSE}
# Genotype file names
SNP_file_name = paste(base.dir, "[genotype file]", sep="")
snps_location_file_name = paste(base.dir, "[SNP location file]", sep="")

# Gene expression file names
expression_file_name = paste(base.dir, "[expression file]", sep="")
gene_location_file_name = paste(base.dir, "[gene location file]", sep="")

# Output file name
output_file_name = paste0(base.dir, "[give your output a filename]")
```

<br><br>Next, we need to decide on the p-value threshold. Only the associations that are significant at this level will be saved in your output file. We often use a p-value threshold of 5 * 10^-8 to indicate if a test is significant at the genome wide level. However, here we choose a slightly less stringent threshold, so you can also have a look at less significant effects.

```{r pvalue, eval=TRUE}
pvOutputThreshold = 5e-7
```

<br><br>Let's load the genotype data in. For this first analysis, we do not need the SNP and gene locations yet. MatrixEQTL comes with a specific way to load in large datasets. We only use a limited list of SNPs, so it is not particularly useful here.

```{r load_geno, eval=FALSE}
snps = SlicedData$new();
snps$fileDelimiter = "\t";      # the TAB character
snps$fileOmitCharacters = "NA"; # denote missing values;
snps$fileSkipRows = 1;          # one row of column labels
snps$fileSkipColumns = 1;       # one column of row labels
snps$fileSliceSize = 2000;      # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name)
```

<br><br>**Q2. For how many SNPs did you just load the genotypes? And for how many people?**  
<br>Hint: Individual names are coded here to start with 'ERR', and genetic variants are identified by their reference SNP (rs) number.

<br><br>We will now load the expression data. As there are many genes in the matrix, the 'slicing' of the data will become useful here.

```{r load_exp, eval=FALSE}
gene = SlicedData$new();
gene$fileDelimiter = "\t";      # the TAB character
gene$fileOmitCharacters = "NA"; # denote missing values;
gene$fileSkipRows = 1;          # one row of column labels
gene$fileSkipColumns = 1;       # one column of row labels
gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name)
```

<br><br>**Q3. For how many genes did you just load the expression values? And for how many people? Are there more samples in the genotype or in the expression dataset?**  

<br><br>We will now run an eQTL mapping analysis for the first time. By naming this model 'model 1', and giving other models different names, you can save results with different parameters. Do not forget to also change the output_file_name if you want to save the results.

```{r model1, eval=FALSE}
model1 = Matrix_eQTL_engine(
  snps = snps,
  gene = gene,
  output_file_name = output_file_name,
  pvOutputThreshold = pvOutputThreshold,
  useModel = useModel, 
  verbose = TRUE,
  pvalue.hist = 100,
  min.pv.by.genesnp = FALSE,
  noFDRsaveMemory = FALSE)
```

<br><br>**Q4. How many eQTLs do you find?**  
<br>Hint: you can find the answer within the `model1` object using `model1$all`

<br><br>The object `model1` contains more interesting information about the eQTL analysis. The input parameters, how long the analysis took, and the eQTL results are all saved here. You can use this info to display the results:
```{r result, eval=FALSE}
cat(model1$time.in.sec);
show(model1$all$eqtls)
```

<br><br>**Q5. How many tests were performed in total?**  
<br><br>**Q6. What percentage of those is significant, according to the threshold that we set earlier?**
<br>Hint: use `model1$neqtls` and the answer from **Q5**

<br><br>In the results that we have created now, the associations between all SNPs and all genes are tested. However, you have learned that  there is a difference between *cis*-eQTLs and *trans*-eQTLs. Therefore, we will now continue to run the model separately SNPs that lie close to the genes, and those that are further away from the gene.  

***
<br><br><br>

## *Cis*- and *trans*-eQTL mapping

<br>The first thing to do, again, is to set the names of the output files. This time you will get two separate files.

```{r output_location, eval=FALSE}
output_file_name_cis = paste(base.dir,"[give your cis output a filename]", sep="")
pvOutputThreshold_trans = paste(base.dir,"[give your trans output a filename]", sep="")
```

<br><br>Next, we will set a P-value threshold for both the *cis*-eQTL and *trans*-eQTL associations. Let's start with **5e-2** and **5e-5**, respectively.

```{r pvalues, eval=FALSE}
# Only associations significant at this level will be saved
pvOutputThreshold_cis = [set p-value];
pvOutputThreshold_trans = [set p-value]
```

<br><br>Read in the files with SNP and gene positions.

```{r positions, eval=FALSE}
snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);
genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE)
```

<br><br>**Q7. What information is provided for each SNP and each gene in these files? How many SNPs and genes are available in these files?**

<br><br>Last, we need to set the maximum distance between a SNP and the gene, that we would still call a *cis-*eQTL. The unit of measurement is basepairs. Set the distance to 1Mb.

```{r distance, eval=FALSE}
cisDist = [fill in distance in base pairs]
```

<br><br>Since we still have the data loaded from the first run, the next step is to simply run the model, this time with output for cis- and trans-eQTLs separately.

```{r model2, eval=FALSE}
model2 = Matrix_eQTL_main(
  snps = snps, 
  gene = gene, 
  output_file_name     = output_file_name_trans,
  pvOutputThreshold     = pvOutputThreshold_trans,
  useModel = useModel, 
  verbose = TRUE, 
  output_file_name.cis = output_file_name_cis,
  pvOutputThreshold.cis = pvOutputThreshold_cis,
  snpspos = snpspos, 
  genepos = genepos,
  cisDist = cisDist,
  pvalue.hist = 100,
  min.pv.by.genesnp = FALSE,
  noFDRsaveMemory = FALSE)
```

<br><br>**Q8. How many *cis* and *trans*-eQTLs did you find?**  
<br><br>**Q9. How many *cis* and *trans* tests were performed?  Where does this difference in scale come from?**  
<br><br>**Q10. What percentage of *cis* and *trans* tests are significant eQTLs?**  
<br><br>**Q11. How many unique SNPs are part of a *cis*eQTL? and of a *trans*-eQTL?** 
<br>Hint: use the functions `length()` and `unique()` for this question

***
<br><br><br>

## Plotting

<br>MatrixEQTL has some in-built plotting functions. You can see the plots by running the following:

```{r plot1, eval=FALSE}
plot(model1)
plot(model2)
```

<br><br>**Q12. What is the type of plot you get out?**

<br><br>To explore the options of these plots, run your last model again, with this change:
```{r codeHelp2, eval=FALSE}
  pvalue.hist = "qqplot"
```

<br><br>**Q13. What plot do you get now? What does this tell you about the p-value cut-off we chose?**

<br><br>We will now make some eQTL boxplots, where we show the actual expression of a gene of interest, separated by genotypes. For this example, we will use the top detected eQTL:<br>
**rs1799987 - ENSG00000160791**

```{r codeHelp4, eval=FALSE}
gene2 <- as.matrix(gene)
gene2 <- as.data.frame(gene2[row.names(gene2) == "ENSG00000160791",])
snps2 <- as.matrix(snps)
snps2 <- as.data.frame(snps2[row.names(snps2) == "rs1799987",])
eqtl <- merge(snps2,gene2,by="row.names")
colnames(eqtl) <- c("sample", "genotype", "expression")
eqtl$genotype <- as.factor(eqtl$genotype)
```

```{r boxplot, eval=FALSE}
library(ggplot2)
ggplot(data=eqtl, aes(x=genotype,y=expression, fill = genotype)) + 
  geom_boxplot() +
  geom_point() +
  theme_bw()
```

<br><br>**Q14. Which genotype causes this gene to be upregulated?**
<br>Hint: genotype 0 here stands for 'GG'. Using online sources like https://www.ncbi.nlm.nih.gov/projects/SNP/, you should be able to figure out what the other allele for this SNP is.

<br><br>**Q15. Make the same plot for the fourth eQTL: rs2058660 - ENSG00000115602**
<br>This time add whether this is a *cis*- or *trans*-eQTL in the title. 
<br>Look up the gene name (the ENSG code) and add that to the title as well.
<br>Hint 1.	find SNP location first (see below)
```{r codeHelpPos, eval = FALSE}
snpspos[snpspos$snpid == "rs2058660"",]
```
<br>Hint 2.	Find gene location in the same way 
<br>Hint 3.	Check positions: is this a local (cis) or distal (trans) effect?
<br>Hint 4.	Add a title by adding this to the command for plotting:
```{r codeHelp6, eval=FALSE}
+ ggtitle("[cis/trans]-eQTL of rs2058660 affecting [gene name]")
```

<br><br>Look rs2058660 up on [GWAS catalog](https://www.ebi.ac.uk/gwas/).

<br><br>**Q16. What phenotype was this SNP first assocatiated with? What was the risk allele? Is there a known link between the GWAS phenotype and the gene?**

***
<br><br><br><br><br><br>
