---
title: "ENHPATHY WORKSHOP GRaNPA"
author: "Nila Servaas"
output:
  html_document:
    toc: yes
  html_notebook:
    toc: yes
    number_sections: no
    toc_float: yes
    theme: spacelab
    code_folding: none
  pdf_document:
    toc: yes
editor_options:
  markdown:
    wrap: 72
---

## References

This vignette is a modified version of [the official GRaNPA workflow vignette](https://grp-zaugg.embl-community.io/GRaNPA/articles/workflow.html).

## Introduction

This workshop vignette shows how to use the GRaNPA package. You will learn in detail how to work with a GRaNPA object and what its main functions and properties are. You will use example data from the GRaNPA package and the macrophage GRN that we constructed in the GRaNIE workshop. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dir <- "/Users/servaas/Documents/EMBL/Workshops/ENHPATHY_2022/GRaNPA/" #replace with the directory where you have your data for the GRaNPA workshop stored.
setwd(dir)
```

## Install and load the required packages to run GRaNPA

In order to run GRaNPA, we have to install the GRaNPA package and its required dependencies. 

Once you have installed a package, you need to load it before you can use it: <br>
`library("package name")`. You can only make use of the package and its functions/variables after installation and loading.

Lets load the required so we can start using them. We use `suppressPackageStartupMessages` here to suppress the output messages of the packages for reasons of brevity. 

```{r lib, eval=TRUE}
#Install devtools from CRAN:
install.packages("devtools")

#Install GRaNPA from the Zaugg group Gitlab using devtools
library(devtools)
devtools::install_gitlab("grp-zaugg/GRaNPA", host = "git.embl.de", force = TRUE)

#Load the needed packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(GRaNIE)
  library(GRaNPA)
})
```

## Loading our GRN

We want to use GRaNPA to evaluate our macrophage GRN that we just made in the GRaNIE workshop. Firstly, we need to load the GRN object and transform into a format that can be accepted by GRaNPA. It should always contain these two columns: `TF.name` and `gene.ENSEMBL`. Since GRaNIE and GRaNPA are well integrated with each other, our GRN object already contains these columns. In case you want to use GRaNPA to evaluate a GRN that was created with another tool than GRaNIE, make sure these columns are present.

```{r Load GRN}
#Load the GRaNIE object
GRN <- readRDS("../GRaNIE/output/GRN.rds")

#Retrieve filtered GRN connections from GRaNIE to evauluate using GRaNPA
GRN_connections.all = getGRNConnections(GRN, type = "all.filtered", include_TF_gene_correlations = TRUE)

#View the GRN connections to make sure it has the columns we need
#These should be ‘TF.name’ and ‘gene.ENSEMBL
GRN_connections.all
```

## Loading a differential expression dataset

In order to run GRaNPA, we also need a differential expression (DE) dataset which will be used to evaluate the GRN. We use the example dataset that is include in the GRaNPA package (`GRaNPA_DE_example`). This dataset is from a study where they compared resting macrophages to macrophages infected with Salmonella bacteria.

The DE dataset should always contain these three columns: `ENSEMBL` (IDs for all genes included in the DE comparison), `padj` (adjusted pvalue for each gene in the DE comparison) and `logFC` (log2 fold change for each gene in the DE comparison). You can input any DE dataset that you like, as long as it contains these columns.

```{r Load DE data}
#Load the DE data from the GRaNPA package
DE = GRaNPA::GRaNPA_DE_example

#View the dataset
DE
```
## Run the GRaNPA predidition

We will now run the main function for GRaNPA. It will train a random forest based on provided GRN to predict the DE values. Here the DE values have been filtered by adjusted pvalue (`DE_pvalue_th`) 0.1 and absolute log2 fold-change (`logFC_th`) 0.58. You can play around with these parameters to see how the prediction changes.

In addition, we can set the number of times that GRaNPA will run the predictions for the real data (`num_run`), and two generated background datasets (`num_run_CR` and `num_run_random`). Since GRaNPA  trains a random forest regression model to predict a differential expression value per gene, the predictions change slightly every time. This is why the predictions need to be ran multiple times so we can estimate an approximate predictive power of our GRN. Running the predicitons at least 3 times is suggested.

For the other parameters, we use the default suggested by GRaNPA. For more information you can check the function manual using `?GRaNPA_main_function`.

```{r Run preditcion}
#If you have a powerful computer, you can set the number of cores here higher (for example 5 or 10)
#By using more cores, the function runes faster.
#If you use two cores, this function takes around 3 minutes to run
GRaNPA_result = GRaNPA::GRaNPA_main_function(DE_data = DE, 
                                               GRN_matrix_filtered = GRN_connections.all,
                                               DE_pvalue_th = 0.1,
                                               logFC_th = 0.58,
                                               num_run = 3,
                                               num_run_CR = 3,
                                               num_run_random = 3,
                                               cores = 2,
                                               importance = "permutation",
                                               ML_type = "regression",
                                               control = "oob")
```

## Checking the prediction results 

After the GRaNPA function has finished, you can plot and check the prediction results.

```{r Plot predicitions}
# Plot the predictions
# The plot will be outputted to the console as well as saved to a pdf
#Replace the output folder with a folder of your choice

outdir = "/Users/servaas/Documents/EMBL/Workshops/ENHPATHY_2022/GRaNPA/"

GRaNPA::plot_GRaNPA_density(GRaNPA.object = GRaNPA_result, plot_name = "GRaNPA_prediction_density.pdf", outputFolder = outdir, width = 4, height = 4)
```
Take a look at the plot outputted by GRaNPA above. On the x-axis, we see the prediction power (R2). We also see different coloured density profiles for the real (yellow) and random networks (green and red). A performance greater than 0.05 (R2) for the real network indicates that the network is doing a good job at prediciting the DE data. A performance greater than 0.05 for a random network to predict random values indicates that the model is overfitting. 

## Finding important TFs for the DE prediciton

Next to prediciting how well our GRN can predict differential gene expression, GRaNPA can also be used to identify the TFs that are important for the prediction. We can use the `plot_GRaNPA_TF_imp` function to see which TFs in our macrophage GRN are important for prediciting macropahge responses to Salmonella infection.

```{r Identify important TFs}
#Plot TF importantce for the top 10 TFs
GRaNPA::plot_GRaNPA_TF_imp(GRaNPA.object = GRaNPA_result, plot_name = "GRaNPA_TF_imp.pdf", num = 10, output = outdir, width = 4, height = 4) 
```

Take a look at the plot outputted by GRaNPA above. On the x-axis, we see scaled importance of every single TF (indicating on the y-axis) for predicting the DE data.

## Wrapping up

Hooray! You are now finished with the main workflow, and have a basic understanding of how to evaluate a GRN using GRaNPA! We encourage you to try out different DE datasets to identify candidate driver TFs for a specific expression response in your GRNs. We hope that the GRaNPA package is useful for your research and encourage you to contact us if you have any question or feature request!
